### ISSUE TITLE
Feature Request: Spatial Transformer Network

### ISSUE THREAD

**Issue by ronrest** [pytorch/pytorch#496]
It would be good to have the components needed to implement Spatial Transformer Networks ( [Jaderberg et al](http://arxiv.org/abs/1506.02025) )

There is an implementation of these components already for [Lua](https://github.com/qassemoquab/stnbh). But it would be good to have these as part of the main package and for Python. 

Tensorflow already has an [implementation of this](https://github.com/tensorflow/models/tree/master/transformer), as does [Lasagne](https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/special.py#L354-L432). There is also an unoficial  [Cafe implementation](https://github.com/daerduoCarey/SpatialTransformerLayer)

**Comment by soumith** [comment]
good call, we'll add this.

**Comment by ronrest** [comment]
Awesome :)

**Comment by wangg12** [comment]
@ronrest any news on the example?

**Comment by fxia22** [comment]
Here is an CPU version of pytorch stn module: https://github.com/fxia22/stn.pytorch
In the meantime I am working on the GPU version of that.

**Comment by jcjohnson** [comment]
I'd also like to see this in main PyTorch. Bilinear sampling in particular is used in recent methods for object detection such as [Mask R-CNN](https://arxiv.org/abs/1703.06870), [Instance-Aware Semantic Segmentation](https://arxiv.org/abs/1512.04412), and [DenseCap](https://arxiv.org/abs/1511.07571).

Also it seems that cuDNN has implemented forward / backward for bilinear sampling; it would be nice if that could be accessible for PyTorch.

I wrote an implementation of bilinear sampling using autograd: https://gist.github.com/jcjohnson/876ca05163ad23ab06c2f98cf3bcd6bb but it uses way too much memory to be much use with large numbers of channels (e.g. 1024 channels for Mask R-CNN with ResNet-50) or large minibatches.

**Comment by soumith** [comment]
okay i'll work on this to be pushed into core.

**Comment by edgarriba** [comment]
@soumith what's the status of bilinear sampling?

**Comment by soumith** [comment]
i'm working on this today.

**Comment by BestSonny** [comment]
@soumith Any update?

**Comment by soumith** [comment]
the first PR with bilinear sampling is now up #2007 
Includes CPU and CuDNN support.

The smaller bit remaining is the grid generator from affine matrix `theta`

**Comment by BestSonny** [comment]
@fxia22 could you make a pull request for the grid generator?

**Comment by fxia22** [comment]
@BestSonny Sure, I will add the affine grid generator only.

**Comment by soumith** [comment]
i'm adding it, on the GPU especially it is efficient to use CuDNN directly (they have functions to generate the grid). I'll have it done in a bit

**Comment by fxia22** [comment]
@soumith I see, cool.

**Comment by BestSonny** [comment]
@soumith Thank you for the excellent STN support. 
In @jcjohnson's DenseCap, the grid and the input can have a different size of height and width, which is a more generalized solution.

Should the current solution support that?

**Comment by soumith** [comment]
no, it doesn't support that, but it's mostly because i added size checks that input == output size. Once I remove that, the code itself is general enough to have a grid that's of a different size than input.

**Comment by BestSonny** [comment]
@soumith I think that would be better.

**Comment by jcjohnson** [comment]
Thanks so much @soumith for putting this together!

I agree that allowing different input and output sizes would be great; for object detection the common case is to take a relatively large feature map for the entire image and produce several smaller crops for objects.

In addition to supporting different input and output sizes, it would also be convenient to allow multiple crops from the same input; however it's probably sufficient to support nonsequential Tensors that have been expanded on the minibatch dimension.

**Comment by soumith** [comment]
I've pushed changes and tests to allow different input and output sizes. 
I've also added tests to make sure that zero-strided inputs are supported.

One thing to keep in mind is that CuDNN's Transformer API does not support zero-strided inputs. Hence, the best I am doing on the GPU side is:
```
if 0 in input.stride():
    input = input.contiguous()
```

If you are not under very high memory pressure, this should be okay (with the caching allocator and all, it should be unnoticeable).

**Comment by jcjohnson** [comment]
Too bad that cuDNN doesn't support zero-strides, but manually forcing contiguous will probably be fine. As a quick back-of-the-envelope, Faster R-CNN with VGG16 will take feature maps roughly of shape 512 x 18 x 25 for each image and produce 300 crops for each input image; expanding the features to 300 x 512 x 18 x 25 is about 264MB. Not ideal but should still be workable.

**Comment by jtoy** [comment]
awesome, is this going out in the next release? looking forward to using this.

**Comment by soumith** [comment]
yes

**Comment by ruotianluo** [comment]
I guess related to discussion above.
https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/vision.py#L27
and
https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/vision.py#L45
are doing contiguous on the same tensor twice. Wouldn't it be a waste of memory?

**Comment by soumith** [comment]
it's not a waste of memory. as soon as Tensor in #L27 goes out of scope, the memory is given back to the allocator.

**Comment by ruotianluo** [comment]
When should it be freed? In principle, if the contiguous one is not saved, it should immediately be released after the grid_sample function, correct?
```
import torch
x = torch.cuda.FloatTensor(1, 1, 16384, 16384)
x = torch.autograd.Variable(x, requires_grad=True)
y = x.expand(2, x.size(1), x.size(2), x.size(3))
grid = torch.rand(2, 1, 1, 2)
import torch.nn.functional as F
z = F.grid_sample(y, torch.autograd.Variable(grid.cuda()))
```
After running the last line, the memory is tripled. Is that as expected?

**Comment by soumith** [comment]
This is expected.
You're again looking at `nvidia-smi` usage. This is not reflective of actual memory usage. The Caching memory allocator built into THC holds the memory blocks for future use.

**Comment by soumith** [comment]
If there is memory pressure, the caching allocator releases the blocks.

**Comment by kamchol89** [comment]
hi 
There was no problem when input size was 48 * 48, but when it was 96 * 48, there was problem like the following figure

Do you know why this is happening?


It's the transformer I used.(https://github.com/oarriaga/STN.keras/blob/master/src/models/layers.py)



![p](https://user-images.githubusercontent.com/8243705/48749280-8cccb480-ecbd-11e8-847e-c529a3e0ba64.JPG)
